blog/_posts/2024-03-05-scala-3.3.2-post-mortem.md

```diff
@@ -0,0 +1,100 @@
+
+In this post, we take a closer look at what happened, why it slipped through our existing testing processes, and what steps we are taking so that it does not happen again in the future.
+
+## Symptoms of the bug
+
+First, let us look at the scenario in which the bug surfaces:
+
+1. Compile a library A with Scala 3.3.2
+2. Compile a project B, depending on A, with Scala 3.3.1 (or 3.3.0)
+3. The older compiler crashes with the following message:
+
+  ```
+  [error] error while loading SomeClass,
+  [error] class file .../SomeClass.class is broken, reading aborted with class java.util.NoSuchElementException
+  [error] contextual$
+  [error] one error found
+  ```
+
+The error does not surface in all situations.
+It happens when `SomeClass`, coming from library A and required to compile B, contains a method whose result type is a context function type.
+Nevertheless, it is not rare.
+
+Since the above scenario is supposed to be accepted according to our forward compatibility guarantees, we had to abort the release of Scala 3.3.2.
+However, the artifacts had already been published to Maven Central, which is an immutable repository.
+We had no choice but to abandon the 3.3.2 version number, and publish 3.3.3 instead.
+
+## The bug in detail
+
+Let us dig a bit deeper on the bug.
+It appears while the compiler is *reading* the TASTy files of a dependency.
+In particular, it cannot read the *name* of a method parameter.
+
+Names in TASTy and in the compiler are not simple Strings.
+They have some amount of structure.
+For example, the name of the class behind an `object Foo` definition has a structured name `ModuleClass("Foo")`, instead of the string `"Foo$"`.
+
+Likewise, generated names for anonymous entites, such as `implicit` parameters, have the structured name `UniqueName("evidence$", 1)`.
+The string `"evidence$"` used in the equality definition of names, so that `UniqueName("evidence$", 1)` is not the same name as `UniqueName("foo$", 1)`, but it does not otherwise have any semantic meaning in TASTy.
+
+However, for code readability and performance reasons, the compiler maintains a fixed list of the prefixes it generates, and uses a single instance of the associated `NameKind`, one of its internal data structures.
+When it reads the string `"evidence$"`, it reuses the instance of `NameKind` that it uses for that particular string.
+
+This leads us to the cause of our bug.
+Between 3.3.1 and 3.3.2, we merged [PR #18280](https://github.com/lampepfl/dotty/pull/18280).
+That PR starts with what appeared to be a *refactoring* commit that renamed some `NameKind`s and their associated strings.
+Unfortunately, that means the 3.3.2 compiler started emitting `UniqueName`s with the string `"contextual$"`.
+When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"contextual$"`, and does not find any, because it did not use to generate that string.
```

Comment:
```suggestion
When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"contextual$"`, and does not find any, because it did not expect to generate that string.
```